<a href="https://colab.research.google.com/github/saurabhdaund76/NLP-/blob/main/NLP.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

So stemming basically converts the words but, some the words transformed in steeming are not meaningful like it converts histori, for final it converts fina etc

In [1]:
import nltk
nltk.download('all')

[nltk_data] Downloading collection 'all'
[nltk_data]    | 
[nltk_data]    | Downloading package abc to /root/nltk_data...
[nltk_data]    |   Unzipping corpora/abc.zip.
[nltk_data]    | Downloading package alpino to /root/nltk_data...
[nltk_data]    |   Unzipping corpora/alpino.zip.
[nltk_data]    | Downloading package biocreative_ppi to
[nltk_data]    |     /root/nltk_data...
[nltk_data]    |   Unzipping corpora/biocreative_ppi.zip.
[nltk_data]    | Downloading package brown to /root/nltk_data...
[nltk_data]    |   Unzipping corpora/brown.zip.
[nltk_data]    | Downloading package brown_tei to /root/nltk_data...
[nltk_data]    |   Unzipping corpora/brown_tei.zip.
[nltk_data]    | Downloading package cess_cat to /root/nltk_data...
[nltk_data]    |   Unzipping corpora/cess_cat.zip.
[nltk_data]    | Downloading package cess_esp to /root/nltk_data...
[nltk_data]    |   Unzipping corpora/cess_esp.zip.
[nltk_data]    | Downloading package chat80 to /root/nltk_data...
[nltk_data]    |   Unzipp

True

In [2]:
nltk.download('punkt')

[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Package punkt is already up-to-date!


True

In [3]:
# nltk is the most imp library in nlp followed by stem 

from nltk.stem import PorterStemmer
from nltk.corpus import stopwords

In [4]:
paragraph = """I have three visions for India. In 3000 years of our history, people from all over 
               the world have come and invaded us, captured our lands, conquered our minds. 
               From Alexander onwards, the Greeks, the Turks, the Moguls, the Portuguese, the British,
               the French, the Dutch, all of them came and looted us, took over what was ours. 
               Yet we have not done this to any other nation. We have not conquered anyone. 
               We have not grabbed their land, their culture, 
               their history and tried to enforce our way of life on them. 
               Why? Because we respect the freedom of others.That is why my 
               first vision is that of freedom. I believe that India got its first vision of 
               this in 1857, when we started the War of Independence. It is this freedom that
               we must protect and nurture and build on. If we are not free, no one will respect us.
               My second vision for India’s development. For fifty years we have been a developing nation.
               It is time we see ourselves as a developed nation. We are among the top 5 nations of the world
               in terms of GDP. We have a 10 percent growth rate in most areas. Our poverty levels are falling.
               Our achievements are being globally recognised today. Yet we lack the self-confidence to
               see ourselves as a developed nation, self-reliant and self-assured. Isn’t this incorrect?
               I have a third vision. India must stand up to the world. Because I believe that unless India 
               stands up to the world, no one will respect us. Only strength respects strength. We must be 
               strong not only as a military power but also as an economic power. Both must go hand-in-hand. 
               My good fortune was to have worked with three great minds. Dr. Vikram Sarabhai of the Dept. of 
               space, Professor Satish Dhawan, who succeeded him and Dr. Brahm Prakash, father of nuclear material.
               I was lucky to have worked with all three of them closely and consider this the great opportunity of my life. 
               I see four milestones in my career"""

In [5]:

sentences = nltk.sent_tokenize(paragraph)
stemmer = PorterStemmer()

In [6]:
# download stop words
nltk.download('stopwords')

[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Package stopwords is already up-to-date!


True

In [7]:
for i in range(len(sentences)):
    words = nltk.word_tokenize(sentences[i])
    words = [stemmer.stem(word) for word in words if word not in set(stopwords.words('english'))]
    sentences[i] = ' '.join(words)  

In [8]:
print(sentences)

['I three vision india .', 'In 3000 year histori , peopl world come invad us , captur land , conquer mind .', 'from alexand onward , greek , turk , mogul , portugues , british , french , dutch , came loot us , took .', 'yet done nation .', 'We conquer anyon .', 'We grab land , cultur , histori tri enforc way life .', 'whi ?', 'becaus respect freedom others.that first vision freedom .', 'I believ india got first vision 1857 , start war independ .', 'It freedom must protect nurtur build .', 'If free , one respect us .', 'My second vision india ’ develop .', 'for fifti year develop nation .', 'It time see develop nation .', 'We among top 5 nation world term gdp .', 'We 10 percent growth rate area .', 'our poverti level fall .', 'our achiev global recognis today .', 'yet lack self-confid see develop nation , self-reli self-assur .', 'isn ’ incorrect ?', 'I third vision .', 'india must stand world .', 'becaus I believ unless india stand world , one respect us .', 'onli strength respect stre

Here we can see that stemming has run successfully and as we discuss some words by stemming do not have any meaning to the words here like histori, peopl,invad etc

**Lemmatization**  it does same word as stemming but lemmatization converts these words into meaningful words

stop words are used because these words dose not play a meaningful role during sentimental analysis

In [9]:
# wordnetlemmitizer is the library important for doing lamitization
import nltk
from nltk.stem import WordNetLemmatizer
from nltk.corpus import stopwords

In [10]:
lemi_sentences = nltk.sent_tokenize(paragraph)
lemmitizer = WordNetLemmatizer()

In [11]:
for i in range(len(lemi_sentences)):
  words = nltk.word_tokenize(lemi_sentences[i]) # it will take the 1st sentence and convert it into words and this loop will run till len(sentence)
  words = [lemmitizer.lemmatize(word) for word in words if word not in set(stopwords.words('english'))] # lemitize word will store all the words except stop words
  lemi_sentences[i] = ' '.join(words) # join the words and convert back into sentences

In [12]:
print(lemi_sentences)

['I three vision India .', 'In 3000 year history , people world come invaded u , captured land , conquered mind .', 'From Alexander onwards , Greeks , Turks , Moguls , Portuguese , British , French , Dutch , came looted u , took .', 'Yet done nation .', 'We conquered anyone .', 'We grabbed land , culture , history tried enforce way life .', 'Why ?', 'Because respect freedom others.That first vision freedom .', 'I believe India got first vision 1857 , started War Independence .', 'It freedom must protect nurture build .', 'If free , one respect u .', 'My second vision India ’ development .', 'For fifty year developing nation .', 'It time see developed nation .', 'We among top 5 nation world term GDP .', 'We 10 percent growth rate area .', 'Our poverty level falling .', 'Our achievement globally recognised today .', 'Yet lack self-confidence see developed nation , self-reliant self-assured .', 'Isn ’ incorrect ?', 'I third vision .', 'India must stand world .', 'Because I believe unless 

Now we can  clearly see the difference that lemi has all the meaning ful words while incase of stemming it was not the case

**TF-IDF (term frequency, inverse documentry frequncy)**

In [13]:
import nltk

In [14]:
paragraph =  """I have three visions for India. In 3000 years of our history, people from all over 
               the world have come and invaded us, captured our lands, conquered our minds. 
               From Alexander onwards, the Greeks, the Turks, the Moguls, the Portuguese, the British,
               the French, the Dutch, all of them came and looted us, took over what was ours. 
               Yet we have not done this to any other nation. We have not conquered anyone. 
               We have not grabbed their land, their culture, 
               their history and tried to enforce our way of life on them. 
               Why? Because we respect the freedom of others.That is why my 
               first vision is that of freedom. I believe that India got its first vision of 
               this in 1857, when we started the War of Independence. It is this freedom that
               we must protect and nurture and build on. If we are not free, no one will respect us.
               My second vision for India’s development. For fifty years we have been a developing nation.
               It is time we see ourselves as a developed nation. We are among the top 5 nations of the world
               in terms of GDP. We have a 10 percent growth rate in most areas. Our poverty levels are falling.
               Our achievements are being globally recognised today. Yet we lack the self-confidence to
               see ourselves as a developed nation, self-reliant and self-assured. Isn’t this incorrect?
               I have a third vision. India must stand up to the world. Because I believe that unless India 
               stands up to the world, no one will respect us. Only strength respects strength. We must be 
               strong not only as a military power but also as an economic power. Both must go hand-in-hand. 
               My good fortune was to have worked with three great minds. Dr. Vikram Sarabhai of the Dept. of 
               space, Professor Satish Dhawan, who succeeded him and Dr. Brahm Prakash, father of nuclear material.
               I was lucky to have worked with all three of them closely and consider this the great opportunity of my life. 
               I see four milestones in my career"""

In [15]:
# cleaning the text

import re # importing regualr expression library
from nltk.corpus import stopwords
from nltk.stem import PorterStemmer
from nltk.stem import WordNetLemmatizer

In [16]:
wordnet = WordNetLemmatizer() # object wordnet to store wordnetlamitizer

In [17]:
# now to convert paragraph into sentences lets use word tokanizer
sentences = nltk.sent_tokenize(paragraph)
sentences
# here we have 31 sentences

['I have three visions for India.',
 'In 3000 years of our history, people from all over \n               the world have come and invaded us, captured our lands, conquered our minds.',
 'From Alexander onwards, the Greeks, the Turks, the Moguls, the Portuguese, the British,\n               the French, the Dutch, all of them came and looted us, took over what was ours.',
 'Yet we have not done this to any other nation.',
 'We have not conquered anyone.',
 'We have not grabbed their land, their culture, \n               their history and tried to enforce our way of life on them.',
 'Why?',
 'Because we respect the freedom of others.That is why my \n               first vision is that of freedom.',
 'I believe that India got its first vision of \n               this in 1857, when we started the War of Independence.',
 'It is this freedom that\n               we must protect and nurture and build on.',
 'If we are not free, no one will respect us.',
 'My second vision for India’s developme

In [18]:
corpus= []
for i in range(len(sentences)):
  review = re.sub('[^a-zA-Z]', ' ', sentences[i])  # here we are replacing all the words except a to z and captical a to z  by simple space and words except this wont play imp role
  review = review.lower()
  review = review.split()
  review = [wordnet.lemmatize(word) for word in review if not word in set(stopwords.words('english'))] # here we lamitized all the words and taken only those except stopwords
  review = ' '.join(review)
  corpus.append(review)

In [19]:
print(sentences)
print(corpus)


['I have three visions for India.', 'In 3000 years of our history, people from all over \n               the world have come and invaded us, captured our lands, conquered our minds.', 'From Alexander onwards, the Greeks, the Turks, the Moguls, the Portuguese, the British,\n               the French, the Dutch, all of them came and looted us, took over what was ours.', 'Yet we have not done this to any other nation.', 'We have not conquered anyone.', 'We have not grabbed their land, their culture, \n               their history and tried to enforce our way of life on them.', 'Why?', 'Because we respect the freedom of others.That is why my \n               first vision is that of freedom.', 'I believe that India got its first vision of \n               this in 1857, when we started the War of Independence.', 'It is this freedom that\n               we must protect and nurture and build on.', 'If we are not free, no one will respect us.', 'My second vision for India’s development.', 'For 

As we can see lamitization has done its job as it removed all unncessary words and cnverted into meaningful words

Now creating tf-idf model

In [20]:
from sklearn.feature_extraction.text import TfidfVectorizer
cv = TfidfVectorizer()
X = cv.fit_transform(corpus).toarray()

In [21]:
X

array([[0.        , 0.        , 0.        , ..., 0.        , 0.        ,
        0.        ],
       [0.        , 0.        , 0.        , ..., 0.25883507, 0.30512561,
        0.        ],
       [0.        , 0.28867513, 0.        , ..., 0.        , 0.        ,
        0.        ],
       ...,
       [0.        , 0.        , 0.        , ..., 0.        , 0.        ,
        0.        ],
       [0.        , 0.        , 0.        , ..., 0.        , 0.        ,
        0.        ],
       [0.        , 0.        , 0.        , ..., 0.        , 0.        ,
        0.        ]])

**Implementing word embeded techniques using embedding layer in keras**

In [28]:
!pip install tensorflow==2.0.0-beta1

     |████████████████████████████████| 88.7 MB 13 kB/s 
     |████████████████████████████████| 3.1 MB 25.1 MB/s 
     |████████████████████████████████| 496 kB 76.0 MB/s 
     |████████████████████████████████| 50 kB 6.8 MB/s 
  Attempting uninstall: tensorflow
    Found existing installation: tensorflow 2.6.0
    Uninstalling tensorflow-2.6.0:
      Successfully uninstalled tensorflow-2.6.0
ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
kapre 0.3.5 requires tensorflow>=2.0.0, but you have tensorflow 2.0.0b1 which is incompatible.


In [29]:
import tensorflow as tf

/usr/local/lib/python3.7/dist-packages/tensorflow/python/framework/dtypes.py:516: FutureWarning: Passing (type, 1) or '1type' as a synonym of type is deprecated; in a future version of numpy, it will be understood as (type, (1,)) / '(1,)type'.
  _np_qint8 = np.dtype([("qint8", np.int8, 1)])
/usr/local/lib/python3.7/dist-packages/tensorflow/python/framework/dtypes.py:517: FutureWarning: Passing (type, 1) or '1type' as a synonym of type is deprecated; in a future version of numpy, it will be understood as (type, (1,)) / '(1,)type'.
  _np_quint8 = np.dtype([("quint8", np.uint8, 1)])
/usr/local/lib/python3.7/dist-packages/tensorflow/python/framework/dtypes.py:518: FutureWarning: Passing (type, 1) or '1type' as a synonym of type is deprecated; in a future version of numpy, it will be understood as (type, (1,)) / '(1,)type'.
  _np_qint16 = np.dtype([("qint16", np.int16, 1)])
/usr/local/lib/python3.7/dist-packages/tensorflow/python/framework/dtypes.py:519: FutureWarning: Passing (type, 1) or 

In [32]:
pip install keras

In [37]:
from tensorflow import keras

In [39]:
keras.preprocessing

<module 'tensorflow.keras.preprocessing' from '/usr/local/lib/python3.7/dist-packages/tensorflow/python/keras/api/_v2/keras/preprocessing/__init__.py'>

In [42]:
from numpy import array
from tensorflow.keras.preprocessing.text import one_hot

In [46]:
### sentences
sentences=[  'the glass of milk',
     'the glass of juice',
     'the cup of tea',
    'I am a good boy',
     'I am a good developer',
     'understand the meaning of words',
     'your videos are good',]

In [47]:
sentences

['the glass of milk',
 'the glass of juice',
 'the cup of tea',
 'I am a good boy',
 'I am a good developer',
 'understand the meaning of words',
 'your videos are good']

In [45]:
# vocabbulry size indecates the size of the dictonary we want to give
voc_size = 10000

**One hot representation**

In [48]:
# this will give the index of every word given in sentences
onehot_index = [one_hot(word,voc_size) for word in sentences]

In [49]:
onehot_index

[[2055, 5971, 3047, 9891],
 [2055, 5971, 3047, 3770],
 [2055, 4762, 3047, 4037],
 [8352, 6130, 5496, 9294, 7738],
 [8352, 6130, 5496, 9294, 494],
 [9504, 2055, 9632, 3047, 137],
 [8828, 5280, 6921, 9294]]

As we can see we got all the indexes of the sentences like the is 2055 glass is 5971 of is 3047 milk is 9891 etc

**Word embedding presentation it is actully a feature representation and we have to pass dimentions ie how many features we want**

In [50]:
from tensorflow.keras.layers import Embedding
from tensorflow.keras.preprocessing.sequence import pad_sequences  # it make sures if all the sentences have same number of words
from tensorflow.keras.models import Sequential

In [51]:
import numpy as np

In [52]:
sentence_length = 8 # this is the parameter passed while running pad_sequences it ensures how much is the maximum length of the sentence we want 

In [54]:
# Suppose we have 4 words and max len is given as 8 then padding=pre will add zeros before if words in sentences are less and post will add aftr sentence
embeded_docs = pad_sequences(sequences=onehot_index , maxlen=sentence_length,padding='pre')
embeded_docs

array([[   0,    0,    0,    0, 2055, 5971, 3047, 9891],
       [   0,    0,    0,    0, 2055, 5971, 3047, 3770],
       [   0,    0,    0,    0, 2055, 4762, 3047, 4037],
       [   0,    0,    0, 8352, 6130, 5496, 9294, 7738],
       [   0,    0,    0, 8352, 6130, 5496, 9294,  494],
       [   0,    0,    0, 9504, 2055, 9632, 3047,  137],
       [   0,    0,    0,    0, 8828, 5280, 6921, 9294]], dtype=int32)

As we can see padding = pre has given 0 before and converted all the sentences into same length 

In [55]:
dim=20 # this represents how many dimentions we want ie features we want in our embeded presentation

In [58]:
model=Sequential()
model.add(Embedding(voc_size, output_dim=dim,input_length=sentence_length))
model.compile('adam','mse')

In [61]:
model.summary()

Model: "sequential_1"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
embedding_1 (Embedding)      (None, 8, 20)             200000    
Total params: 200,000
Trainable params: 200,000
Non-trainable params: 0
_________________________________________________________________


In [65]:
predicted_vectors= model.predict(embeded_docs)
# this is how our above index values are converted into vector form

In [66]:
# lets see the comaprisns 
print(embeded_docs[0])
print(predicted_vectors[0])

[   0    0    0    0 2055 5971 3047 9891]
[[-4.4142317e-02  3.9209496e-02 -4.7749963e-02  2.9337917e-02
  -1.3447583e-02  4.9766008e-02  1.7115403e-02 -2.2452697e-03
   4.6421815e-02  4.8562910e-02 -3.1358171e-02  4.6953928e-02
  -1.3893843e-04  5.9716403e-06 -3.2344628e-02 -2.6327133e-02
   4.2076577e-02  3.2155368e-02 -8.9856163e-03 -1.9426299e-02]
 [-4.4142317e-02  3.9209496e-02 -4.7749963e-02  2.9337917e-02
  -1.3447583e-02  4.9766008e-02  1.7115403e-02 -2.2452697e-03
   4.6421815e-02  4.8562910e-02 -3.1358171e-02  4.6953928e-02
  -1.3893843e-04  5.9716403e-06 -3.2344628e-02 -2.6327133e-02
   4.2076577e-02  3.2155368e-02 -8.9856163e-03 -1.9426299e-02]
 [-4.4142317e-02  3.9209496e-02 -4.7749963e-02  2.9337917e-02
  -1.3447583e-02  4.9766008e-02  1.7115403e-02 -2.2452697e-03
   4.6421815e-02  4.8562910e-02 -3.1358171e-02  4.6953928e-02
  -1.3893843e-04  5.9716403e-06 -3.2344628e-02 -2.6327133e-02
   4.2076577e-02  3.2155368e-02 -8.9856163e-03 -1.9426299e-02]
 [-4.4142317e-02  3.92094

This indicates that our 1st zero got converted into dimention of 20 vectors similarly for others, so above is our 1st sentence 